### PR TITLE
Hyundai longitudinal (CAN): don't send FCA11 on non-FCA cars

### PR DESCRIPTION
**Pull request by sunnyhaibin**
**Prerequisite of**
- https://github.com/commaai/openpilot/pull/26967

### CODE REVIEW

**Comment by sshane** [comment]
Do you know any cases where this happens to already-supported vehicles?

File: selfdrive/car/hyundai/hyundaican.py [hunk 1]
```diff
@@ -119,6 +121,12 @@ def create_acc_commands(packer, enabled, accel, upper_jerk, idx, lead_visible, s
     "aReqValue": accel,  # stock ramps up and down respecting jerk limit until it reaches aReqRaw
     "CR_VSM_Alive": idx % 0xF,
   }
+
+  # send CF_VSM_ConfMode and AEB_Status to cars that do not accept FCA messages, otherwise TCS13|ACCEnable would fault
```
Comment (sshane) [inline selfdrive/car/hyundai/hyundaican.py:125]: can you make this comment more clear? not sending these signals won't make it fault, sending the FCA messages below would. (this is to show AEB disabled in the dash right?)

File: selfdrive/car/hyundai/hyundaican.py [hunk 2]
```diff
@@ -119,6 +121,12 @@ def create_acc_commands(packer, enabled, accel, upper_jerk, idx, lead_visible, s
...
+  if not has_fca:
+    scc12_values["CF_VSM_ConfMode"] = 1
```
Comment (sshane) [inline selfdrive/car/hyundai/hyundaican.py:127]: know what `CF_VSM_ConfMode` does?

**Comment by sunnyhaibin** [comment]
> Do you know any cases where this happens to already-supported vehicles?

I'm not too sure, I will need to review the list of cars that do not have FCA messages and verify if they were able to use openpilot longitudinal or not.

**Comment by sshane** [comment]
Ah I can check the data if you don't know of any cases. Won't be for a while though

File: selfdrive/car/hyundai/hyundaican.py (hunk 2 above)
Comment (sunnyhaibin) [reply selfdrive/car/hyundai/hyundaican.py:127]: It very likely stands for "Cruise Function | Vehicle Stability Management | Configuration Mode", but as far as what each bit means, we did not have a chance to test that. When the stock SCC system is in use, this signal stays at `2`.

**Comment by sunnyhaibin** [comment]
> Ah I can check the data if you don't know of any cases immediately. Won't be for a bit though as I have other stuff with higher priority.

Sounds good, I will try to back track a bit when I have some time freed up as well.

File: selfdrive/car/hyundai/hyundaican.py (hunk 1 above)
Comment (sunnyhaibin) [reply selfdrive/car/hyundai/hyundaican.py:125]: Here are the scenarios we tested:
1. Not sending either - Fault ❌
2. Send `CF_VSM_ConfMode = 1` only - Fault ❌
3. Send `AEB_Status = 1` only - Fault ❌
4. Send `CF_VSM_ConfMode = 1` and `AEB_Status = 1` - No fault, AEB amber light stays on ✅

File: selfdrive/car/hyundai/hyundaican.py (hunk 2 above)
Comment (sshane) [reply selfdrive/car/hyundai/hyundaican.py:127]: What led you to believe you should set it to 2?

Comment (sunnyhaibin) [reply selfdrive/car/hyundai/hyundaican.py:127]: We tried setting this to 2 along with sending `AEB_Status = 1`, but it resulted in a fault.

Comment (sshane) [reply selfdrive/car/hyundai/hyundaican.py:127]: ~What led you to believe you should set it to 1?~ @gregjhogan do you know how VSM plays with FCA/AEB? I see we can request braking via FCA, AEB or VSM

File: selfdrive/car/hyundai/hyundaican.py (hunk 1 above)
Comment (sshane) [reply selfdrive/car/hyundai/hyundaican.py:125]: Hmm, so you weren't able to forward the message through without a fault?

**Comment by sunnyhaibin** [comment]
Backtracking to this PR. Here are the ones that don't have FCA messages but have SCC12 and are able to use openpilot longitudinal without this PR at the moment:
- `IONIQ_PHEV_2019`
- `KIA_NIRO_PHEV`
- `GENESIS_G90`

**Comment by sunnyhaibin** [comment]
@sshane Backtracking to this PR. Here are the ones that don't have FCA messages but have SCC12 and are able to use openpilot longitudinal without this PR at the moment:
- `IONIQ_PHEV_2019`
- `KIA_NIRO_PHEV`
- `GENESIS_G90`

**Comment by adeebshihadeh** [comment]
@sshane I think we should be able to detect this with test_models

**Comment by sshane** [comment]
I tried to do this, but there will be false positives. FCA/SCC isn't sent when radar is disabled, so every disabled radar route gets flagged, as well as sending tester present.

File: selfdrive/car/hyundai/hyundaican.py [hunk 3]
```diff
@@ -187,14 +198,17 @@ def create_acc_opt(packer):
...
-    "FCA_USM": 1, # AEB disabled
-  }
-  commands.append(packer.make_can_msg("FCA12", 0, fca12_values))
+  # For cars that do not broadcast FCA messages, do not send FCA messages to prevent TCS13|ACCEnable fault
+  if has_fca:
+    fca12_values = {
+      "FCA_DrvSetState": 2,
+      "FCA_USM": 1,  # AEB disabled
+    }
+    commands.append(packer.make_can_msg("FCA12", 0, fca12_values))
```
Comment (sshane) [inline selfdrive/car/hyundai/hyundaican.py:207]: Checked some routes, FCA11 and FCA12 existing don't necessarily equal. FCA11 can be on the bus, but not FCA12 often.

**Comment by sshane** [comment]
60k routes last 180 days:

```python
bool tuple is (fca11_exists, fca12_exists), and dongles
KIA STINGER GT2 2018                 - {(True, False): 8, (False, False): 1}
HYUNDAI PALISADE 2020                - {(True, True): 56, (False, False): 11}
HYUNDAI SONATA 2020                  - {(True, True): 57, (False, False): 20}
HYUNDAI SONATA 2019                  - {(False, False): 5}
HYUNDAI SANTA FE 2019                - {(True, True): 25, (False, False): 10}
HYUNDAI SANTA FE 2022                - {(True, True): 12, (False, False): 3}
HYUNDAI SANTA FE HYBRID 2022         - {(True, True): 5, (False, False): 1}
HYUNDAI SONATA HYBRID 2021           - {(True, True): 19, (False, False): 5}
GENESIS G70 2018                     - {(True, False): 1}
KIA NIRO EV 2020                     - {(False, False): 13, (True, False): 28}
KIA OPTIMA 4TH GEN FACELIFT          - {(True, False): 12, (False, False): 1}
HYUNDAI KONA ELECTRIC 2019           - {(True, False): 5}
KIA STINGER 2022                     - {(True, True): 2, (False, False): 1}
HYUNDAI SANTA FE PlUG-IN HYBRID 2022 - {(True, True): 5}
GENESIS G80 2017                     - {(False, False): 5}
HYUNDAI GENESIS 2015-2016            - {(False, False): 7}
KIA K5 2021                          - {(False, False): 2, (True, True): 3}
HYUNDAI ELANTRA 2021                 - {(False, False): 6, (True, True): 17}
HYUNDAI IONIQ PHEV 2020              - {(True, True): 8}
HYUNDAI ELANTRA HYBRID 2021          - {(False, False): 6, (True, True): 9}
HYUNDAI IONIQ HYBRID 2020-2022       - {(True, True): 1}
KIA SORENTO GT LINE 2018             - {(False, False): 3}
KIA NIRO HYBRID 2021                 - {(True, True): 2}
HYUNDAI ELANTRA 2017                 - {(True, False): 3}
KIA SELTOS 2021                      - {(True, True): 1, (False, False): 1}
HYUNDAI IONIQ ELECTRIC LIMITED 2019  - {(False, False): 1}
KIA FORTE E 2018 & GT 2021           - {(True, True): 1}
KIA K5 HYBRID 2020                   - {(True, True): 1}
HYUNDAI IONIQ HYBRID 2017-2019       - {(False, False): 3}
GENESIS G70 2020                     - {(True, False): 3}
HYUNDAI IONIQ ELECTRIC 2020          - {(False, False): 1, (True, True): 1}
KIA NIRO HYBRID 2019                 - {(False, False): 2}
HYUNDAI KONA HYBRID 2020             - {(True, True): 2}
KIA CEED INTRO ED 2019               - {(True, True): 1}
HYUNDAI TUCSON 2019                  - {(True, False): 1}
HYUNDAI KONA ELECTRIC 2022           - {(True, True): 1}
```

**Comment by sshane** [comment]
Need to test these cases to ensure the AEB light still shows:
- [ ] FCA11: yes, FCA12: no
- [ ] FCA11: no, FCA12: no (set the signal in SCC12, check on one of those cars)

**Comment by sshane** [comment]
Forgot to exclude segments when `openpilotLongitudinalControl`, posted updated list above. Looks like all the cases where FCA11 does not exist, the car is behind the legacy safety mode, so this is a safe change!

File: selfdrive/car/hyundai/hyundaican.py [hunk 4]
```diff
@@ -141,6 +141,7 @@ def create_acc_commands(packer, enabled, accel, upper_jerk, idx, lead_visible, s
     }
   commands.append(packer.make_can_msg("SCC11", 0, scc11_values))
 
+  # TODO: add AEB_Status?
```
Comment (sshane) [inline selfdrive/car/hyundai/hyundaican.py:144]:
```suggestion
```

File: selfdrive/car/hyundai/hyundaican.py [hunk 5]
```diff
@@ -187,6 +190,7 @@ def create_acc_opt(packer):
   }
   commands.append(packer.make_can_msg("SCC13", 0, scc13_values))
 
+  # TODO: ensure this is enough to show the light on cars without FCA11
```
Comment (sshane) [inline selfdrive/car/hyundai/hyundaican.py:193]:
```suggestion
```

**Comment by sshane** [comment]
thanks! you can open a new PR for whatever else you need to do to support for legacy cars

**Comment by sunnyhaibin** [comment]
> thanks! you can open a new PR for whatever else you need to do to support for legacy cars

Cool!

**Pull request merged**